qualtran/bloqs/qsp_bloq.py

```diff
@@ -0,0 +1,149 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Any, Dict, Sequence
+
+import cirq
+import numpy as np
+from attrs import frozen
+from cirq_ft import GateWithRegisters, Register, Signature
+from numpy.typing import NDArray
+
+
+def _arbitrary_SU2_rotation(theta: float, phi: float, lambd: float):
+    r"""Implements an arbitrary SU(2) rotation defined by
+
+    .. math::
+
+        \begin{matrix}
+        e^{i(\lambda + \phi)} \cos(\theta) & e^{i\phi} \sin(\theta) \\
+        e^{i\phi} \sin(\theta) & - \cos(\theta)
+        \end{matrix}
+
+    Returns:
+        A 2x2 rotation matrix
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Equation 7.
+    """
+    return np.array(
+        [
+            [np.exp(1j * (lambd + phi)) * np.cos(theta), np.exp(1j * phi) * np.sin(theta)],
+            [np.exp(1j * lambd) * np.sin(theta), -np.cos(theta)],
+        ]
+    )
+
+
+def qsp_phase_factors(P: Sequence[float], Q: Sequence[float]) -> Dict[str, Any]:
+    """Computes the QSP signal rotations for a given pair of polynomials.
+    The QSP transformation is described in Theorem 3, and the algorithm for computing co-efficients is described in Algorithm 1.
+
+    Args:
+        P: Co-efficients of a complex polynomial.
+        Q: Co-efficients of a complex polynomial.
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 3; Algorithm 1.
+    """
+    if len(P) != len(Q):
+        raise ValueError("Polynomials P and Q must have the same degree.")
+
+    S = np.array([P, Q])
+    n = S.shape[1]
+
+    theta = np.zeros(n)
+    phi = np.zeros(n)
+    lambd = 0
+
+    for d in range(n - 1, 0, -1):
+        assert S.shape == (2, d + 1)
+
+        a, b = S[:, d]
+        theta[d] = np.arctan2(np.abs(a), np.abs(b))
+        phi[d] = np.angle(a / b)
+
+        if d == 0:
+            lambd = np.angle(b)
+        else:
+            S = _arbitrary_SU2_rotation(theta[d], phi[d], 0) @ S
+            S = np.array([S[0][1:d], S[1][0 : d - 1]])
+
+    return {'theta': theta, 'phi': phi, 'lambda': lambd}
+
+
+@frozen
+class QEVTCircuit(GateWithRegisters):
+    r"""Applies a QSP sequence described by a pair of polynomials $P, Q$, to a unitary $U$ to obtain a block-encoding of $P(U)$.
+    The exact circuit is described in Figure 2.
```

Comment:
nit: Each line in the docstring should be < 100 characters. Also, the summary line should be a single line followed by an empty line followed by a detailed description. 

See https://peps.python.org/pep-0257/#multi-line-docstrings